api/src/main/java/io/minio/Signer.java

```diff
@@ -142,16 +145,15 @@ private void setCanonicalHeaders() {
     this.signedHeaders = Joiner.on(";").join(this.canonicalHeaders.keySet());
   }
 
-
   private void setCanonicalQueryString() {
-    Map<String,String> signedQueryParams = new TreeMap<>();
-
     String encodedQuery = this.url.encodedQuery();
     if (encodedQuery == null) {
       this.canonicalQueryString = "";
       return;
     }
 
+    Multimap<String, String> signedQueryParams = MultimapBuilder.treeKeys().arrayListValues().build();
```

Comment:
Just curious to know the difference with `Multimap<String, String> signedQueryParams = TreeMultimap.create();`

Reply:
> Just curious to know the difference with `Multimap<String, String> signedQueryParams = TreeMultimap.create();` Even though it works like this, `TreeMultimap.create()` is quite straight forward.

TreeMultimap orders keys & values, meaning if one key has several values, values will also be ordered.

Though this what S3 spec says, Minio server doesn't support it, so I need to send other PRs in Minio server/minio-go/minio-java and merge them at the same time and release all of them together to fix this.

Reply:
This means it will fail working with AWS S3 till it's fixed correctly.  IMO minio server issue needs to be handled separately than fixing minio-java to work with minio server's issue.

Reply:
> This means it will fail working with AWS S3 till it's fixed correctly. IMO minio server issue needs to be handled separately than fixing minio-java to work with minio server's issue.

This is a specific Minio feature, AWS S3 doesn't support it. Then fixing Minio server will make codes using minio-go deployment fail, that's why I wanted more coordination about this

Reply:
My point here is, it is Signer where the bug is introduced.  Sending single or multiple values to query parameters are higher level behaviours, but Signer needs to work correct for requests.

Reply:
> My point here is, it is Signer where the bug is introduced. Sending single or multiple values to query parameters are higher level behaviours, but Signer needs to work correct for requests.

yes, I agree. I am just saying it is a TODO thing (we are actually not making any regressions here) and I prefer if this is done with some coordination to avoid having users complaining.